File: bindings/dotnet/src/Zlink.Codecs.Protobuf/ProtoCodec.cs
// SPDX-License-Identifier: MPL-2.0

using Google.Protobuf;
using Zlink;

namespace Zlink.Codecs.Protobuf;

public static class ProtoCodec
{
    public static T Decode<T>(this Message message, MessageParser<T> parser)
        where T : IMessage<T>
    {
        ArgumentNullException.ThrowIfNull(message);
        ArgumentNullException.ThrowIfNull(parser);
        return parser.ParseFrom(message.AsReadOnlySpan());
    }

    public static Message Encode<T>(this T value)
        where T : IMessage<T>
    {
        ArgumentNullException.ThrowIfNull(value);
        return Message.FromOwnedBytes(value.ToByteArray());
    }
}

File: bindings/java/codec/zlink-codec-protobuf/src/main/java/dev/kairoscode/zlink/codec/protobuf/ProtoCodec.java
package dev.kairoscode.zlink.codec.protobuf;

import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Parser;
import java.util.Objects;

final class ProtoCodec<T extends com.google.protobuf.Message>
    implements MessageCodec<T> {
    private final Parser<T> parser;

    public ProtoCodec(Parser<T> parser) {
        this.parser = Objects.requireNonNull(parser, "parser");
    }

    @Override
    public dev.kairoscode.zlink.Message toMessage(T value) {
        Objects.requireNonNull(value, "value");
        return dev.kairoscode.zlink.Message.copyOf(value.toByteArray());
    }

    @Override
    public T fromMessage(dev.kairoscode.zlink.Message message) {
        Objects.requireNonNull(message, "message");
        try {
            return parser.parseFrom(message.toByteArray());
        } catch (InvalidProtocolBufferException ex) {
            throw new IllegalArgumentException(
                "failed to decode protobuf payload", ex);
        }
    }
}
